### ISSUE TITLE
Suggestion: don't render if `permalink: false` and `eleventyExcludeFromCollections: true`

### ISSUE THREAD

**Issue by nhoizey** [11ty/eleventy#1700]
While trying to identify the source of issue #1696, I discovered that templates/layouts are evaluated even if their `permalink` is set to `false`, either directly in the source file Front Matter or via the data cascade, including `_data/eleventyComputed.js`.

Here's a reduced test case, in the `computed-permalink-false` branch of this repository:
https://github.com/nhoizey/11ty-reduced-test-cases/tree/computed-permalink-false

If the template/layout tries to use `{{ page.outputPath }}` or `{{ page.url }}` with filters built for strings, there will be an error, as these have a `false` value.

For example, I had to replace `{% if 'archives' in page.url %}` with `{% if page.url != false and 'archives' in page.url %}` because `{% if 'archives' in page.url %}` gives an error when `{{ page.url }}` equals `false`.

As no page will be generated, it might be better not to parse the content of the source file, apart from the Front Matter, once the data cascade has been completed, if `permalink` is evaluated to `false`.

It might even make the full build faster.

**Comment by nhoizey** [comment]
An alternative could be to set `{{ page.outputPath }}` and `{{ page.url }}` anyway, even if no file is written in the end, to prevent issues when using them in templates with filters.

**Comment by edwardhorsford** [comment]
I've run in to this myself - having to use eleventyIgnore in a really manual way to get Eleventy to stop evaluating a file. I'd much prefer if I could do this on the basis of frontmatter - which could be related to the requested feature of supporting drafts or a callback / frontmatter / function to determine whether to process a file.

**Comment by nhoizey** [comment]
@zachleat that's right, I should have mentioned that I also set `eleventyExcludeFromCollections` to `true` when I set `permalink` to `false` for drafts during production builds.

So maybe what I suggest here could be done if both `eleventyExcludeFromCollections === true` AND `permalink === false`?

**Comment by zachleat** [comment]
Ah hmm, that’s a fair idea!

I do want to point out in the mean time that serverless permalink objects solve this problem already: https://www.11ty.dev/docs/plugins/serverless/#step-3-use-a-permalink-object

Templates only render with a classic `permalink` or a permalink object with a `build` key.

**Comment by zachleat** [comment]
> This repository is now using [lodash style issue management](https://twitter.com/samselikoff/status/991395669016436736) for enhancements. This means enhancement issues will now be closed instead of leaving them open.

> View [the enhancement backlog here](https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+). Don’t forget to upvote the top comment with 👍!

**Comment by andy-bond** [comment]
@zachleat thanks for your comment above about how to *really, really* disable outputting the file. 

I, [like a few others](https://github.com/11ty/eleventy/issues/2060), wanted a "draft" mechanism, but ran into issues with `permalink: false` causing issues in unexpected places because it was still present when doing things like `{% if item.link in page.url %}`.

I ended up with the following, based on [@pdehaan's example here.](https://github.com/pdehaan/11ty-drafts/blob/main/src/src.11tydata.js):
```js
module.exports = {
	layout: 'base',
	tags: 'documentation',
	eleventyComputed: {
		permalink: (data) => {
			if (data.draft && process.env.NODE_ENV === 'production') {
				return {}; // <-- really, really remove from build
			}
			return data.permalink !== '' ? data.permalink : `/${data.page.fileSlug}/`;
		},
		eleventyExcludeFromCollections: (data) => {
			if (data.draft && process.env.NODE_ENV === 'production') {
				return true;
			}
			return data.eleventyExcludeFromCollections;
		},
	},
};
```

**Comment by zachleat** [comment]
Preprocessors is the recommended path forward for this one! (shipped with v3.0.0)

#188 

https://www.11ty.dev/docs/config-preprocessors/#example-drafts
